File: src/treemapper/diffctx/edges/semantic/ansible.py
from __future__ import annotations

import re
from pathlib import Path

from ...config.weights import EDGE_WEIGHTS
from ...types import Fragment
from ..base import EdgeBuilder, EdgeDict, FragmentIndex, discover_files_by_refs

_ANSIBLE_EXTS = {".yml", ".yaml"}
_TEMPLATE_EXTS = {".j2", ".jinja2", ".jinja"}

_INCLUDE_VARS_RE = re.compile(
    r"^\s*(?:include_vars|vars_files)\s*:\s*[\"']?([^\s\"']{1,300})[\"']?",
    re.MULTILINE,
)
_INCLUDE_TASKS_RE = re.compile(
    r"^\s*(?:include_tasks|import_tasks|include_role|import_role|import_playbook)" r"\s*:\s*[\"']?([^\s\"']{1,300})[\"']?",
    re.MULTILINE,
)
_TEMPLATE_SRC_RE = re.compile(
    r"^\s*src\s*:\s*[\"']?([^\s\"']{1,300}\.j(?:2|inja2?))[\"']?",
    re.MULTILINE,
)
_ROLES_LIST_RE = re.compile(
    r"^\s*-\s+(?:role:\s*)?([a-zA-Z_][\w.-]{0,200})\s*$",
    re.MULTILINE,
)
_VARS_FILES_LIST_RE = re.compile(
    r"^\s*-\s+[\"']([^\"']{1,300}\.ya?ml)[\"']",
    re.MULTILINE,
)


def _is_ansible_file(path: Path) -> bool:
    return path.suffix.lower() in _ANSIBLE_EXTS or path.suffix.lower() in _TEMPLATE_EXTS


def _get_role_name_from_path(path: Path) -> str | None:
    parts = path.parts
    for i, part in enumerate(parts):
        if part == "roles" and i + 1 < len(parts):
            return parts[i + 1]
    return None


def _extract_refs(content: str, file_path: Path) -> set[str]:
    refs: set[str] = set()

    for m in _INCLUDE_VARS_RE.finditer(content):
        refs.add(m.group(1))
    for m in _INCLUDE_TASKS_RE.finditer(content):
        refs.add(m.group(1))
    for m in _TEMPLATE_SRC_RE.finditer(content):
        refs.add(m.group(1))
    for m in _VARS_FILES_LIST_RE.finditer(content):
        refs.add(m.group(1))

    role_name = _get_role_name_from_path(file_path)
    if role_name and "/tasks/" in str(file_path):
        refs.add(f"roles/{role_name}/handlers/main.yml")
        refs.add(f"roles/{role_name}/templates/")
        refs.add(f"roles/{role_name}/files/")

    return refs


def _extract_role_refs(content: str) -> set[str]:
    roles: set[str] = set()
    in_roles_block = False
    for line in content.splitlines():
        stripped = line.strip()
        if stripped.startswith("roles:"):
            in_roles_block = True
            continue
        if in_roles_block:
            if stripped.startswith("- "):
                m = _ROLES_LIST_RE.match(line)
                if m:
                    roles.add(m.group(1))
            elif stripped and not stripped.startswith("#"):
                if not stripped.startswith("-") and ":" in stripped:
                    in_roles_block = False
    return roles


def _ref_to_filename(ref: str) -> str:
    name = ref.rstrip("/").split("/")[-1].lower()
    return name


class AnsibleEdgeBuilder(EdgeBuilder):
    weight = 0.60
    include_weight = EDGE_WEIGHTS["ansible_include"].forward
    role_weight = EDGE_WEIGHTS["ansible_role"].forward
    reverse_weight_factor = EDGE_WEIGHTS["ansible_include"].reverse_factor

    def discover_related_files(
        self,
        changed_files: list[Path],
        all_candidate_files: list[Path],
        repo_root: Path | None = None,
    ) -> list[Path]:
        ansible_changed = [f for f in changed_files if _is_ansible_file(f)]
        if not ansible_changed:
            return []

        refs: set[str] = set()
        role_names: set[str] = set()

        for f in ansible_changed:
            try:
                content = f.read_text(encoding="utf-8")
                for ref in _extract_refs(content, f):
                    refs.add(_ref_to_filename(ref))
                    refs.add(ref)
                for role in _extract_role_refs(content):
                    role_names.add(role)
            except (OSError, UnicodeDecodeError):
                continue

            role_name = _get_role_name_from_path(f)
            if role_name:
                role_names.add(role_name)

        for role in role_names:
            refs.add(f"roles/{role}/tasks/main.yml")
            refs.add(f"roles/{role}/handlers/main.yml")
            refs.add(f"roles/{role}/templates/")

        self._add_group_vars_refs(ansible_changed, all_candidate_files, refs, repo_root)

        return discover_files_by_refs(refs, changed_files, all_candidate_files, repo_root)

    def _add_group_vars_refs(
        self,
        ansible_changed: list[Path],
        all_candidate_files: list[Path],
        refs: set[str],
        _repo_root: Path | None,
    ) -> None:
        for f in ansible_changed:
            rel = str(f)
            if "group_vars/" in rel or "host_vars/" in rel:
                for candidate in all_candidate_files:
                    try:
                        content = candidate.read_text(encoding="utf-8")
                        fname = f.name
                        if fname in content or str(f) in content:
                            refs.add(candidate.name.lower())
                    except (OSError, UnicodeDecodeError):
                        continue
            try:
                content = f.read_text(encoding="utf-8")
                if "vars_files:" in content or "include_vars" in content:
                    for m in _VARS_FILES_LIST_RE.finditer(content):
                        refs.add(m.group(1))
                    for m in _INCLUDE_VARS_RE.finditer(content):
                        refs.add(m.group(1))
            except (OSError, UnicodeDecodeError):
                continue

    def build(self, fragments: list[Fragment], repo_root: Path | None = None) -> EdgeDict:
        ansible_frags = [f for f in fragments if _is_ansible_file(f.path)]
        if not ansible_frags:
            return {}

        edges: EdgeDict = {}
        idx = FragmentIndex(fragments, repo_root)

        for af in ansible_frags:
            self._add_fragment_edges(af, idx, edges)

        self._link_role_siblings(ansible_frags, edges)

        return edges

    def _add_fragment_edges(self, af: Fragment, idx: FragmentIndex, edges: EdgeDict) -> None:
        refs = _extract_refs(af.content, af.path)
        for ref in refs:
            self.link_by_name_or_path(af.id, ref, idx, edges, self.include_weight, filename=_ref_to_filename(ref))

        roles = _extract_role_refs(af.content)
        for role in roles:
            for subdir in ("tasks", "handlers", "templates"):
                path_hint = f"roles/{role}/{subdir}"
                self.link_by_path_match(af.id, path_hint, idx, edges, self.role_weight)

    def _link_role_siblings(self, ansible_frags: list[Fragment], edges: EdgeDict) -> None:
        role_frags: dict[str, list[Fragment]] = {}
        for f in ansible_frags:
            role = _get_role_name_from_path(f.path)
            if role:
                role_frags.setdefault(role, []).append(f)

        sibling_weight = self.weight * 0.6
        for _, frags in role_frags.items():
            for i, f1 in enumerate(frags):
                for f2 in frags[i + 1 :]:
                    self.add_edge(edges, f1.id, f2.id, sibling_weight)

File: diffctx/src/edges/semantic/ansible.rs
use std::path::{Path, PathBuf};

use once_cell::sync::Lazy;
use regex::Regex;
use rustc_hash::{FxHashMap, FxHashSet};

use crate::config::edge_weights::ANSIBLE_SEMANTIC;
use crate::config::weights::EDGE_WEIGHTS;
use crate::types::Fragment;

use super::super::EdgeDict;
use super::super::base::{
    self, EdgeBuilder, FragmentIndex, add_edge, discover_files_by_refs, link_by_name,
    link_by_path_match,
};

static ANSIBLE_EXTS: Lazy<FxHashSet<&str>> = Lazy::new(|| {
    [".yml", ".yaml", ".j2", ".jinja2", ".jinja"]
        .iter()
        .copied()
        .collect()
});

static INCLUDE_VARS_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"(?m)^\s*(?:include_vars|vars_files)\s*:\s*["']?([^\s"']{1,300})["']?"#).unwrap()
});
static INCLUDE_TASKS_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"(?m)^\s*(?:include_tasks|import_tasks|include_role|import_role|import_playbook)\s*:\s*["']?([^\s"']{1,300})["']?"#).unwrap()
});
static TEMPLATE_SRC_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"(?m)^\s*src\s*:\s*["']?([^\s"']{1,300}\.j(?:2|inja2?))["']?"#).unwrap()
});
static ROLES_LIST_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^\s*-\s+(?:role:\s*)?([a-zA-Z_][\w.\-]{0,200})\s*$").unwrap());
static VARS_FILES_LIST_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r#"(?m)^\s*-\s+["']([^"']{1,300}\.ya?ml)["']"#).unwrap());

fn is_ansible_file(path: &Path) -> bool {
    let ext = base::file_ext(path);
    ANSIBLE_EXTS.contains(ext.as_str())
}

fn get_role_name(path: &Path) -> Option<String> {
    let parts: Vec<&str> = path.iter().filter_map(|c| c.to_str()).collect();
    for (i, part) in parts.iter().enumerate() {
        if *part == "roles" && i + 1 < parts.len() {
            return Some(parts[i + 1].to_string());
        }
    }
    None
}

fn ref_to_filename(r: &str) -> String {
    r.trim_end_matches('/')
        .split('/')
        .next_back()
        .unwrap_or(r)
        .to_lowercase()
}

fn extract_refs(content: &str, file_path: &Path) -> FxHashSet<String> {
    let mut refs = FxHashSet::default();
    for m in INCLUDE_VARS_RE.captures_iter(content) {
        refs.insert(m[1].to_string());
    }
    for m in INCLUDE_TASKS_RE.captures_iter(content) {
        refs.insert(m[1].to_string());
    }
    for m in TEMPLATE_SRC_RE.captures_iter(content) {
        refs.insert(m[1].to_string());
    }
    for m in VARS_FILES_LIST_RE.captures_iter(content) {
        refs.insert(m[1].to_string());
    }

    if let Some(role) = get_role_name(file_path) {
        let path_str = file_path.to_string_lossy();
        if path_str.contains("/tasks/") {
            refs.insert(format!("roles/{}/handlers/main.yml", role));
            refs.insert(format!("roles/{}/templates/", role));
            refs.insert(format!("roles/{}/files/", role));
        }
    }
    refs
}

fn extract_role_refs(content: &str) -> FxHashSet<String> {
    let mut roles = FxHashSet::default();
    let mut in_roles = false;
    for line in content.lines() {
        let stripped = line.trim();
        if stripped.starts_with("roles:") {
            in_roles = true;
            continue;
        }
        if !in_roles {
            continue;
        }
        if stripped.starts_with("- ") {
            if let Some(c) = ROLES_LIST_RE.captures(line) {
                roles.insert(c[1].to_string());
            }
        } else if !stripped.is_empty() && !stripped.starts_with('#') && stripped.contains(':') {
            in_roles = false;
        }
    }
    roles
}

pub struct AnsibleEdgeBuilder;

impl EdgeBuilder for AnsibleEdgeBuilder {
    fn build(&self, fragments: &[Fragment], repo_root: Option<&Path>) -> EdgeDict {
        let frags: Vec<&Fragment> = fragments
            .iter()
            .filter(|f| is_ansible_file(Path::new(f.path())))
            .collect();
        if frags.is_empty() {
            return FxHashMap::default();
        }

        let include_w = EDGE_WEIGHTS["ansible_include"].forward;
        let role_w = EDGE_WEIGHTS["ansible_role"].forward;
        let rev = EDGE_WEIGHTS["ansible_include"].reverse_factor;
        let sibling_w = include_w * ANSIBLE_SEMANTIC.sibling_modifier;

        let idx = FragmentIndex::new(fragments, repo_root);
        let mut edges: EdgeDict = FxHashMap::default();

        for af in &frags {
            let path = Path::new(af.path());
            for r in extract_refs(&af.content, path) {
                let filename = ref_to_filename(&r);
                link_by_name(&af.id, &filename, &idx, &mut edges, include_w, rev);
                link_by_path_match(&af.id, &r, &idx, &mut edges, include_w, rev);
            }

            for role in extract_role_refs(&af.content) {
                for subdir in ["tasks", "handlers", "templates"] {
                    let path_hint = format!("roles/{}/{}", role, subdir);
                    link_by_path_match(&af.id, &path_hint, &idx, &mut edges, role_w, rev);
                }
            }
        }

        let mut role_frags: FxHashMap<String, Vec<&Fragment>> = FxHashMap::default();
        for f in &frags {
            if let Some(role) = get_role_name(Path::new(f.path())) {
                role_frags.entry(role).or_default().push(f);
            }
        }
        for group in role_frags.values() {
            for (i, f1) in group.iter().enumerate() {
                for f2 in &group[i + 1..] {
                    add_edge(&mut edges, &f1.id, &f2.id, sibling_w, rev);
                }
            }
        }

        edges
    }

    fn discover_related_files(
        &self,
        changed: &[PathBuf],
        candidates: &[PathBuf],
        repo_root: Option<&Path>,
        file_cache: Option<&FxHashMap<PathBuf, String>>,
    ) -> Vec<PathBuf> {
        let ansible_changed: Vec<&PathBuf> =
            changed.iter().filter(|p| is_ansible_file(p)).collect();
        if ansible_changed.is_empty() {
            return vec![];
        }

        let mut refs = FxHashSet::default();
        let mut role_names = FxHashSet::default();

        for f in &ansible_changed {
            if let Some(content) = base::read_file_cached(f, file_cache) {
                for r in extract_refs(&content, f) {
                    refs.insert(ref_to_filename(&r));
                    refs.insert(r);
                }
                for role in extract_role_refs(&content) {
                    role_names.insert(role);
                }
            }
            if let Some(role) = get_role_name(f) {
                role_names.insert(role);
            }
        }

        for role in &role_names {
            refs.insert(format!("roles/{}/tasks/main.yml", role));
            refs.insert(format!("roles/{}/handlers/main.yml", role));
            refs.insert(format!("roles/{}/templates/", role));
        }

        discover_files_by_refs(&refs, changed, candidates, repo_root)
    }

    fn category_label(&self) -> Option<&str> {
        Some("semantic")
    }
}
